cranelift/codegen/src/egraph.rs

```diff
@@ -255,6 +284,23 @@ where
             optimized_values.len()
         );
 
+        optimized_values.push(orig_value);
+
+        let mut best_block = self.available_block[*optimized_values.last().unwrap()];
```

Comment:
This loop looks correct to me (very slick in-place compaction!) but could we have a comment describing what it's doing? It took me a little bit, with knowledge of what it's supposed to be doing, to grok it. Perhaps something like: "Remove any values from `optimized_values` that do not have the highest possible available block in the domtree." at the top, and "This value's available block is higher; all higher indices we've already scanned have lower blocks and need to be removed, so truncate after this element." in the first case, and "This value's available block is lower, and we want to skip it; remove it from the list and keep going (order does not matter so `swap_remove` is OK)." in the second?

Reply:
One of the things I'm not sure here is that, when `swap_remove()` is called, the element that was at the end of `optimized_values` is put in the `idx-th` position, but then it's never used again to lookup `this_block` for the next iteration.  Is this the desirable behavior, or am I missing something here?

Reply:
(Re: your edit) yep, indeed; I think it might be good to add a comment with the loop invariant here, for clarity. Something like: "Loop scans in reverse; all values at indices >= idx have the same available block, which is the best (highest in domtree) available block seen so far."